File: src/sopmclient.cpp
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>

#include "sopmclient.h"

#define BUFFER_SIZE 256

SoPMClient::SoPMClient() : isConnected_(false), sockfd_(socket(AF_INET, SOCK_STREAM, 0))
{

}

SoPMClient::~SoPMClient()
{
    closeConnection();
}

void SoPMClient::connectToServer(std::string address, int port)
{
    struct sockaddr_in socketAddress;
    int result;

    socketAddress.sin_family = AF_INET;
    socketAddress.sin_addr.s_addr = inet_addr(address.c_str());
    socketAddress.sin_port = htons(port);

    result = connect(sockfd_, (struct sockaddr *)&socketAddress, sizeof(socketAddress));
    if (result == -1) {
        throw "Failed connection to the server";
    }
    isConnected_ = true;
}

void SoPMClient::closeConnection()
{
    if (!isConnected_) {
        return;
    }
    close(sockfd_);
    isConnected_ = false;
}

bool SoPMClient::hasConnection()
{
    return isConnected_;
}

std::string SoPMClient::request(std::string request)
{
    if (isConnected_ == false) {
        throw "No connection";
    }

    char buffer[BUFFER_SIZE];
    int result;

    snprintf(buffer, BUFFER_SIZE, "%s", request.c_str());
    if (result < 0) {
        throw "Failed sending request to th server";
    }

    result = recv(sockfd_, buffer, BUFFER_SIZE, 0);
    if (result <= 0) {
        throw "Failed receiveing answer from the server";
    }

    return buffer;
}

File: src/main/java/SoPMClient.java
import java.io.*;
import java.net.Socket;

public class SoPMClient {
    private static final String IP_ADDRESS = "localhost";
    private static final int PORT = 5000;

    private static boolean isConnected_;
    private static Socket socket_;

    public SoPMClient(){
        isConnected_ = false;
    }

    public static void connectToServer() throws IOException {
        try {
            socket_ = new Socket(IP_ADDRESS,  PORT);
            isConnected_ = true;
        } catch (IOException e) {
            throw e;
        }
    }

    public static void closeConnection() {
        if (!isConnected_) {
            return;
        }
        try {
            socket_.close();
            System.out.println("Connection is closed");
        } catch (IOException e) { /* cannot happen */ }
        isConnected_ = false;
    }

    public static boolean hasConnection() {
        return isConnected_;
    }

    public static String request(String request) {
        try {
            //BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            DataOutputStream out = new DataOutputStream(socket_.getOutputStream());
            System.out.println("Sending request to the ServerSocket");
            out.writeUTF(request);
            System.out.println("Server Wrote message to client.");
            out.flush();

            DataInputStream in = new DataInputStream(socket_.getInputStream());
            System.out.println("DataInputStream created");
            System.out.println("Server reading from channel");
            String entry = in.readUTF();
            System.out.println("READ from client message - " + entry);
            System.out.println("Server try writing to channel");

            return entry;
        }
        catch (IOException e) {
            return "Loser";
        }
    }
}